File: tests/c/test_range.c
/* Copyright (C) 2015, WSID   */

/* Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <glib.h>

#include "crankbase.h"

//////// Declaration ///////////////////////////////////////////////////////////

static void		test_uint_equal (void);

static void		test_uint_hash (void);

static void		test_uint_to_string (void);

static void		test_uint_is_empty (void);

static void		test_uint_is_unit (void);

static void		test_uint_length (void);

static void		test_uint_contains (void);

static void		test_uint_get (void);

static void		test_uint_index_of (void);

static void		test_uint_clamp (void);

static void		test_uint_intersection (void);

//////// Main //////////////////////////////////////////////////////////////////

gint
main (	gint   argc,
      	gchar *argv[] )
{
	g_test_init (&argc, &argv, NULL);

	g_test_add_func ("/crank/base/range/uint/equal",		test_uint_equal	);
	g_test_add_func ("/crank/base/range/uint/hash",			test_uint_hash	);
	g_test_add_func ("/crank/base/range/uint/to_string",	test_uint_to_string	);
	g_test_add_func ("/crank/base/range/uint/is_empty",		test_uint_is_empty	);
	g_test_add_func ("/crank/base/range/uint/is_unit",		test_uint_is_unit	);
	g_test_add_func ("/crank/base/range/uint/length",		test_uint_length	);
	g_test_add_func ("/crank/base/range/uint/contains",		test_uint_contains	);
	g_test_add_func ("/crank/base/range/uint/get",			test_uint_get	);
	g_test_add_func ("/crank/base/range/uint/index_of",		test_uint_index_of	);
	g_test_add_func ("/crank/base/range/uint/clamp",		test_uint_clamp	);
	g_test_add_func ("/crank/base/range/uint/intersection",	test_uint_intersection	);
	
	g_test_run ();

	return 0;
}


//////// Definition ////////////////////////////////////////////////////////////

static void
test_uint_equal ( void )
{
	CrankRanUint	a = {0, 4};
	CrankRanUint	b = {3, 5};
	CrankRanUint	c = {0, 4};
	
	g_assert (! crank_ran_uint_equal (&a, &b));
	g_assert (crank_ran_uint_equal (&a, &c));
}

static void
test_uint_hash ( void )
{
	CrankRanUint	a = {0, 4};
	CrankRanUint	b = {3, 5};
	CrankRanUint	c = {0, 4};
	
	g_assert_cmpuint (crank_ran_uint_hash (&a), !=, crank_ran_uint_hash (&b));
	g_assert_cmpuint (crank_ran_uint_hash (&a), ==, crank_ran_uint_hash (&c));
}

static void
test_uint_to_string (void)
{
	CrankRanUint	a = {3, 5};
	gchar*			astr = crank_ran_uint_to_string (&a);
	
	g_assert_cmpstr ("[3, 5)", ==, astr);
	
	g_free (astr);
}

static void
test_uint_is_empty (void)
{
	CrankRanUint	a = {3, 4};
	g_assert (! crank_ran_uint_is_empty (&a));
	
	crank_ran_uint_init (&a, 6, 6);
	g_assert (crank_ran_uint_is_empty (&a));
}

static void
test_uint_is_unit (void)
{
	CrankRanUint	a = {3, 4};
	g_assert (crank_ran_uint_is_unit (&a));
	
	crank_ran_uint_init (&a, 6, 6);
	g_assert (! crank_ran_uint_is_unit (&a));
}

static void
test_uint_length (void)
{
	CrankRanUint	a;
	
	crank_ran_uint_init (&a, 3, 5);
	
	g_assert_cmpuint (crank_ran_uint_get_length (&a), ==, 2);
}

static void
test_uint_contains (void)
{
	CrankRanUint	a;
	
	crank_ran_uint_init (&a, 4, 8);
	
	g_assert (crank_ran_uint_contains (&a, 4));
	g_assert (crank_ran_uint_contains (&a, 7));
	
	g_assert (! crank_ran_uint_contains (&a, 3));
	g_assert (! crank_ran_uint_contains (&a, 8));
}

static void
test_uint_get (void)
{
	CrankRanUint	a;
	
	crank_ran_uint_init (&a, 10, 110);
	
	g_assert_cmpuint (crank_ran_uint_get (&a, 0.25f), ==, 35);
}

static void
test_uint_index_of (void)
{
	CrankRanUint	a;
	crank_ran_uint_init (&a, 10, 110);
	
	crank_assert_cmpfloat (crank_ran_uint_index_of (&a, 60), ==, 0.5f);
}

static void
test_uint_clamp (void)
{
	CrankRanUint	a;
	crank_ran_uint_init (&a, 55, 95);
	
	g_assert_cmpuint (crank_ran_uint_clamp (&a, 43), ==, 55);
	g_assert_cmpuint (crank_ran_uint_clamp (&a, 77), ==, 77);
	g_assert_cmpuint (crank_ran_uint_clamp (&a, 99), ==, 95);
}

static void
test_uint_intersection (void)
{
	CrankRanUint	a;
	CrankRanUint	b;
	CrankRanUint	c;
	
	CrankRanUint	d;
	CrankRanUint	e;
	
	crank_ran_uint_init (&a, 20, 35);
	crank_ran_uint_init (&b, 45, 80);
	crank_ran_uint_init (&c, 30, 60);
	
	g_assert (! crank_ran_uint_intersection (&a, &b, &d));
	g_assert (crank_ran_uint_intersection (&a, &c, &e));
	
	g_assert_cmpuint (e.start, ==, 30);
	g_assert_cmpuint (e.end, ==, 35);
}

File: tests/python/base/test_range.py
#!/usr/bin/python

# Copyright (C) 2015, WSID

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

import unittest


from gi.repository import GObject
from gi.repository import CrankBase


class TestRange(unittest.TestCase):

	def test_uint_equal (self):
		a = CrankBase.RanUint.init (3, 5)
		b = CrankBase.RanUint.init (0, 4)
		c = CrankBase.RanUint.init (3, 5)
		
		assert (not a.equal (b))
		assert (a.equal (c))
		
	def test_uint_hash (self):
		a = CrankBase.RanUint.init (3, 5)
		b = CrankBase.RanUint.init (0, 4)
		c = CrankBase.RanUint.init (3, 5)
		
		self.assertNotEqual (a.hash (), b.hash ())
		self.assertEqual (a.hash (), c.hash ());

	def test_uint_to_string (self):
		a = CrankBase.RanUint.init (3, 5)
		
		self.assertEqual (a.to_string (), "[3, 5)")
	
	def test_uint_is_empty (self):
		a = CrankBase.RanUint.init (2, 3)
		assert (not a.is_empty())
		
		a = CrankBase.RanUint.init (4, 4)
		assert (a.is_empty())
	
	def test_uint_is_unit (self):
		a = CrankBase.RanUint.init (2, 3)
		assert (a.is_unit())
		
		a = CrankBase.RanUint.init (4, 4)
		assert (not a.is_unit())
		
	def test_uint_get_length (self):
		a = CrankBase.RanUint.init (7, 12)
		self.assertEqual (a.get_length (), 5)

	def test_uint_contains (self):
		a = CrankBase.RanUint.init (10, 23)
		assert (a.contains (15))
		assert (not a.contains (55))

	def test_uint_get (self):
		a = CrankBase.RanUint.init (10, 110)
		self.assertEqual (a.get (0.25), 35)

	def test_uint_index_of (self):
		a = CrankBase.RanUint.init (10, 110)
		self.assertEqual (a.index_of (85), 0.75)
	
	def test_uint_clamp (self):
		a = CrankBase.RanUint.init (10, 100)
		self.assertEqual (a.clamp (4), 10)
		self.assertEqual (a.clamp (44), 44)
		self.assertEqual (a.clamp (200), 100)
	
	def test_uint_intersection (self):
		a = CrankBase.RanUint.init (14, 45)
		b = CrankBase.RanUint.init (55, 74)
		c = CrankBase.RanUint.init (30, 60)
		
		(res, d) = a.intersection (b)
		
		assert (not res)
		
		(res, e) = a.intersection (c)
		
		assert (res)
		self.assertEqual (e.start, 30)
		self.assertEqual (e.end, 45)
		
if __name__ == '__main__':
	unittest.main ()
